File: iconics-java/src/main/java/com/open/iconics/view/IconicsCompoundButton.java
package com.open.iconics.view;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.CompoundButton;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

import com.open.iconics.Iconics;
import com.open.iconics.IconicsDrawable;
import com.open.iconics.internal.CheckableIconBundle;
import com.open.iconics.internal.IconicsView;
import com.open.iconics.internal.IconicsViewsAttrsApplier;
import com.open.iconics.internal.IconicsViewsUtils;

import static androidx.annotation.RestrictTo.Scope.LIBRARY_GROUP;

public class IconicsCompoundButton extends CompoundButton implements IconicsView {
    private final CheckableIconBundle mIconsBundle = new CheckableIconBundle();

    public IconicsCompoundButton(Context context) {
        super(context);
    }

    public IconicsCompoundButton(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public IconicsCompoundButton(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        if (!isInEditMode()) {
            initialize(context, attrs, defStyle);
        }
    }

    @Override
    @RestrictTo(LIBRARY_GROUP)
    public void initialize(Context context, AttributeSet attrs, int defStyle) {
        mIconsBundle.createIcons(context);
        applyAttr(context, attrs, defStyle);
        IconicsViewsUtils.checkAnimation(mIconsBundle.mCheckedIcon, this);
        IconicsViewsUtils.checkAnimation(mIconsBundle.mUncheckedIcon, this);
        setButtonDrawable(mIconsBundle.createStates(context));
    }

    @Override
    @RestrictTo(LIBRARY_GROUP)
    public void applyAttr(Context context, AttributeSet attrs, int defStyle) {
        IconicsViewsAttrsApplier.readIconicsCompoundButton(context, attrs, mIconsBundle);
        mIconsBundle.mAnimateChanges = IconicsViewsAttrsApplier.isIconicsAnimateChanges(context, attrs);
    }

    public void setCheckedIcon(@Nullable IconicsDrawable icon) {
        mIconsBundle.mCheckedIcon = IconicsViewsUtils.checkAnimation(icon, this);
        setButtonDrawable(mIconsBundle.createStates(getContext()));
    }

    public void setUncheckedIcon(@Nullable IconicsDrawable icon) {
        mIconsBundle.mUncheckedIcon = IconicsViewsUtils.checkAnimation(icon, this);
        setButtonDrawable(mIconsBundle.createStates(getContext()));
    }

    @Nullable
    public IconicsDrawable getCheckedIcon() {
        return mIconsBundle.mCheckedIcon;
    }

    @Nullable
    public IconicsDrawable getUncheckedIcon() {
        return mIconsBundle.mUncheckedIcon;
    }

    @Override
    public void setText(@Nullable CharSequence text, @NonNull BufferType type) {
        if (!isInEditMode()) {
            super.setText(
                    new Iconics.IconicsBuilder()
                            .ctx(getContext())
                            .on(text == null ? "" : text)
                            .build(),
                    type);
        } else {
            super.setText(text, type);
        }
    }

    @Override
    public CharSequence getAccessibilityClassName() {
        return IconicsCompoundButton.class.getName();
    }
}

File: iconics-kotlin/src/main/java/com/open/iconics/view/IconicsCompoundButton.kt
package com.open.iconics.view

import android.content.Context
import android.util.AttributeSet
import android.widget.CompoundButton
import android.widget.TextView
import androidx.core.widget.CompoundButtonCompat
import com.open.iconics.IconicsDrawable
import com.open.iconics.animation.tryToEnableIconicsAnimation
import com.open.iconics.internal.CheckableIconBundle
import com.open.iconics.internal.IconicsViewsAttrsApplier
import com.open.iconics.utils.buildIconics

open class IconicsCompoundButton @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : CompoundButton(context, attrs, defStyle) {
    private val iconsBundle = CheckableIconBundle()

    var checkedIcon: IconicsDrawable?
        get() = iconsBundle.checkedIcon
        set(icon) {
            iconsBundle.checkedIcon = tryToEnableIconicsAnimation(icon)
            buttonDrawable = iconsBundle.createStates(context)
        }

    var uncheckedIcon: IconicsDrawable?
        get() = iconsBundle.uncheckedIcon
        set(icon) {
            iconsBundle.uncheckedIcon = tryToEnableIconicsAnimation(icon)
            buttonDrawable = iconsBundle.createStates(context)
        }

    init {
        iconsBundle.createIcons(context)

        IconicsViewsAttrsApplier.readIconicsCompoundButton(context, attrs, iconsBundle)
        iconsBundle.animateChanges = IconicsViewsAttrsApplier.isIconicsAnimateChanges(context, attrs)

        tryToEnableIconicsAnimation(iconsBundle.checkedIcon, iconsBundle.uncheckedIcon)
        buttonDrawable = iconsBundle.createStates(context)

        @Suppress("LeakingThis")
        CompoundButtonCompat.getButtonDrawable(this)?.let {
            minWidth = it.minimumWidth
            minHeight = it.minimumHeight
        }
    }

    override fun setText(text: CharSequence, type: TextView.BufferType) {
        if (!isInEditMode) {
            super.setText(text.buildIconics(), type)
        } else {
            super.setText(text, type)
        }
    }

    override fun getAccessibilityClassName(): CharSequence = IconicsCompoundButton::class.java.name
}
